File: app/src/main/java/com/marknkamau/justjava/ui/login/LogInActivityPresenter.java
package com.marknkamau.justjava.ui.login;

import android.content.SharedPreferences;
import android.os.Handler;

import com.marknkamau.justjava.utils.Constants;
import com.marknkamau.justjava.utils.FirebaseAuthUtils;
import com.marknkamau.justjava.utils.FirebaseDBUtil;

class LogInActivityPresenter {
    private LogInActivityView activityView;
    private SharedPreferences sharedPreferences;

    LogInActivityPresenter(LogInActivityView activityView, SharedPreferences sharedPreferences) {
        this.activityView = activityView;
        this.sharedPreferences = sharedPreferences;
    }

    void checkSignInStatus() {
        if (FirebaseAuthUtils.INSTANCE.getCurrentUser() != null) {
            activityView.closeActivity();
        }
    }

    void signIn(String email, String password) {
        activityView.showDialog();
        FirebaseAuthUtils.INSTANCE.signIn(email, password, new FirebaseAuthUtils.AuthActionListener() {
            @Override
            public void actionSuccessful(String response) {
                getUserDefaults();
            }

            @Override
            public void actionFailed(String response) {
                activityView.displayMessage(response);
            }
        });
    }

    private void getUserDefaults() {
        FirebaseDBUtil.INSTANCE.getUserDefaults(new FirebaseDBUtil.UserDetailsListener() {
            @Override
            public void taskSuccessful(String name, String phone, String deliveryAddress) {
                saveToSharedPreferences(name, phone, deliveryAddress);
                activityView.dismissDialog();
                activityView.displayMessage("Sign in successful");
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        activityView.finishSignUp();
                    }
                }, 500);
            }

            @Override
            public void taskFailed(String response) {
                activityView.displayMessage(response);
            }
        });
    }

    void resetUserPassword(String email) {
        FirebaseAuthUtils.INSTANCE.sendPasswordResetEmail(email, new FirebaseAuthUtils.AuthActionListener() {
            @Override
            public void actionSuccessful(String response) {
                activityView.displayMessage(response);
            }

            @Override
            public void actionFailed(String response) {
                activityView.displayMessage(response);
            }
        });
    }
    private void saveToSharedPreferences(String name, String phone, String address) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(Constants.INSTANCE.getDEF_NAME(), name);
        editor.putString(Constants.INSTANCE.getDEF_PHONE(), phone);
        editor.putString(Constants.INSTANCE.getDEF_ADDRESS(), address);

        editor.apply();
    }
}

File: app/src/main/java/com/marknkamau/justjava/ui/login/LogInActivityPresenter.kt
package com.marknkamau.justjava.ui.login

import android.content.SharedPreferences
import android.os.Handler

import com.marknkamau.justjava.utils.Constants
import com.marknkamau.justjava.utils.FirebaseAuthUtils
import com.marknkamau.justjava.utils.FirebaseDBUtil

internal class LogInActivityPresenter(private val activityView: LogInActivityView, private val sharedPreferences: SharedPreferences) {

    fun checkSignInStatus() {
        if (FirebaseAuthUtils.currentUser != null) {
            activityView.closeActivity()
        }
    }

    fun signIn(email: String, password: String) {
        activityView.showDialog()
        FirebaseAuthUtils.signIn(email, password, object : FirebaseAuthUtils.AuthActionListener {
            override fun actionSuccessful(response: String) {
                getUserDefaults()
            }

            override fun actionFailed(response: String?) {
                activityView.displayMessage(response)
            }
        })
    }

    private fun getUserDefaults() {
        FirebaseDBUtil.getUserDefaults(object : FirebaseDBUtil.UserDetailsListener {
            override fun taskSuccessful(name: String, phone: String, deliveryAddress: String) {
                saveToSharedPreferences(name, phone, deliveryAddress)
                activityView.dismissDialog()
                activityView.displayMessage("Sign in successful")
                Handler().postDelayed({ activityView.finishSignUp() }, 500)
            }

            override fun taskFailed(reason: String?) {
                activityView.displayMessage(reason)
            }
        })
    }

    fun resetUserPassword(email: String) {
        FirebaseAuthUtils.sendPasswordResetEmail(email, object : FirebaseAuthUtils.AuthActionListener {
            override fun actionSuccessful(response: String) {
                activityView.displayMessage(response)
            }

            override fun actionFailed(response: String?) {
                activityView.displayMessage(response)
            }
        })
    }

    private fun saveToSharedPreferences(name: String, phone: String, address: String) {
        val editor = sharedPreferences.edit()
        editor.putString(Constants.DEF_NAME, name)
        editor.putString(Constants.DEF_PHONE, phone)
        editor.putString(Constants.DEF_ADDRESS, address)

        editor.apply()
    }
}
